File: src/main/kotlin/io/titandata/titan/commands/Context.kt
/*
 * Copyright The Titan Project Contributors.
 */

package io.titandata.titan.commands

import com.github.ajalt.clikt.core.CliktCommand
import com.github.ajalt.clikt.core.requireObject
import com.github.ajalt.clikt.core.subcommands
import com.github.ajalt.clikt.parameters.arguments.argument
import com.github.ajalt.clikt.parameters.arguments.optional
import com.github.ajalt.clikt.parameters.options.default
import com.github.ajalt.clikt.parameters.options.flag
import com.github.ajalt.clikt.parameters.options.multiple
import com.github.ajalt.clikt.parameters.options.option
import io.titandata.titan.Dependencies
import io.titandata.titan.exceptions.InvalidArgumentException
import org.kodein.di.Kodein
import org.kodein.di.generic.bind
import org.kodein.di.generic.inSet
import org.kodein.di.generic.provider

class Context : CliktCommand(help = "Manage titan contexts") {
    override fun run() {}
}

class ContextInstall : CliktCommand(help = "Install a new context", name = "install") {
    private val dependencies: Dependencies by requireObject()
    private val type: String by option("-t", "--type", help = "Context type (docker or kubernetes), defaults to docker").default("docker")
    private val name: String? by option("-n", "--name", help = "Context name, defaults to context type")
    private val parameters by option("-p", "--parameters", help = "Context specific parameters. key=value format").multiple()
    private val verbose by option("-v", "--verbose", help = "Verbose logging").flag(default = false)

    override fun run() {
        val contextName = name ?: type
        val params = mutableMapOf<String, String>()
        for (param in parameters) {
            val split = param.split("=")
            if (split.count() != 2) {
                throw InvalidArgumentException(message = "Parameters must be in key=value format.", exitCode = 1, output = param)
            }
            params[split[0]] = split[1]
        }
        val provider = dependencies.providers.create(contextName, type)
        provider.install(params, verbose)
        dependencies.providers.addProvider(provider)
    }
}

class ContextUninstall : CliktCommand(help = "Uninstall a context", name = "uninstall") {
    private val force by option("-f", "--force", help = "Destroy all repositories").flag(default = false)
    private val dependencies: Dependencies by requireObject()
    private val contextName by argument()

    override fun run() {
        val provider = dependencies.providers.byName(contextName)
        provider.uninstall(force, false)
        dependencies.providers.removeProvider(contextName)
    }
}

class ContextList : CliktCommand(help = "List available contexts", name = "ls") {
    private val dependencies: Dependencies by requireObject()
    private val n = System.lineSeparator()

    override fun run() {
        System.out.printf("%-20s  %-12s$n", "NAME", "TYPE")
        val providers = dependencies.providers.list()
        if (!providers.isEmpty()) {
            val defaultName = dependencies.providers.defaultName()
            for (providerEntry in providers) {
                var context = providerEntry.key
                if (providerEntry.key == defaultName) {
                    context += " (*)"
                }
                val type = providerEntry.value.getType()
                System.out.printf("%-20s  %-12s$n", context, type)
            }
        }
    }
}
class ContextDefault : CliktCommand(help = "Get or set default context", name = "default") {
    private val dependencies: Dependencies by requireObject()
    private val contextName by argument().optional()

    override fun run() {
        if (contextName != null) {
            dependencies.providers.setDefault(contextName!!)
        } else {
            println(dependencies.providers.defaultName())
        }
    }
}

val contextModule = Kodein.Module("context") {
    bind<CliktCommand>().inSet() with provider {
        Context().subcommands(
                ContextInstall(),
                ContextUninstall(),
                ContextDefault(),
                ContextList()
        )
    }
}

File: internal/app/commands/context.go
/*
Copyright © 2019 The Titan Project Contributors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
package commands

import (
	"fmt"
	"github.com/spf13/cobra"
	"os"
	"titan/internal/app/providers"
)

var contextType string
var contextName string

// contextCmd represents the context command
var contextCmd = &cobra.Command{
	Use:   "context",
	Short: "Manage titan contexts",
}

// contextInstallCmd represents the contextInstall command
var contextInstallCmd = &cobra.Command{
	Use:   "install",
	Short: "Install a new context",
	Run: func(cmd *cobra.Command, args []string) {
		provider = providers.Create(contextName, contextType, providers.GetAvailablePort())
		provider.Install(nil, verbose) //TODO add params
		providers.AddProvider(provider)
	},
}

// contextUninstallCmd represents the contextUninstall command
var contextUninstallCmd = &cobra.Command{
	Use:   "uninstall [CONTEXTNAME]",
	Short: "Uninstall a context",
	Args: cobra.ExactArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		provider, context := providers.ByName(args[0])
		provider.Uninstall(force, false)
		providers.Remove(context)
	},
}

// contextDefaultCmd represents the contextDefault command
var contextDefaultCmd = &cobra.Command{
	Use:   "default [CONTEXTNAME]",
	Short: "Get or set default context",
	Run: func(cmd *cobra.Command, args []string) {
		if len(args) == 0 {
			p := providers.Default()
			fmt.Println(p.GetName())
			os.Exit(0)
		}
		context = args[0]
		providers.SetDefault(context)
	},
}

// contextListCmd represents the contextList command
var contextListCmd = &cobra.Command{
	Use:   "ls",
	Short: "List available contexts",
	Run: func(cmd *cobra.Command, args []string) {
		h := fmt.Sprintf("%-20s  %-12s", "NAME", "TYPE")
		fmt.Println(h)
		plist := providers.List()
		if len(plist) > 0 {
			defaultName := providers.DefaultName()
			for context, provider := range plist {
				if context == defaultName {
					context = context + " (*)"
				}
				l := fmt.Sprintf("%-20s  %-12s", context, provider.GetType())
				fmt.Println(l)
			}
		}
	},
}

func init() {
	rootCmd.AddCommand(contextCmd)
	contextCmd.AddCommand(contextInstallCmd)
	contextCmd.AddCommand(contextUninstallCmd)
	contextCmd.AddCommand(contextDefaultCmd)
	contextCmd.AddCommand(contextListCmd)

	contextInstallCmd.Flags().StringVarP(&contextType, "type", "t", "docker", "context type (docker or kubernetes)")
	contextInstallCmd.Flags().StringVarP(&contextName, "name", "n", "docker", "context name, defaults to context type")
	contextInstallCmd.Flags().StringSliceVarP(&params, "parameters", "p", nil, "context specific parameters. key=value format")
	contextInstallCmd.Flags().BoolVarP(&verbose, "verbose", "v", false, "verbose logging")
	contextInstallCmd.Flags().SortFlags = false //TODO review flag sorting

	contextUninstallCmd.Flags().BoolVarP(&force, "force", "f", false, "destroy all repositories")

}
